[libc++][doc] Updates module build instructions.

mordante opened the pull request:
CMake has landed experimental support for using the Standard modules. This will be part of the CMake 3.30 release. This updates the build instructions to use modules with CMake.

The changes have been tested locally.

mordante committed:
[lib++][doc] Updates module build instructions.

CMake has landed experimental support for using the Standard modules. This
will be part of the CMake 3.30 release. This updates the build
instructions to use modules with CMake.

The changes have been tested locally.

mordante commented on libcxx/docs/Modules.rst:
```diff
@@ -214,6 +219,62 @@ Building this project is done with the following steps, assuming the files
+
+  set(CMAKE_CXX_STANDARD 23)
+  set(CMAKE_CXX_STANDARD_REQUIRED YES)
+  # This seems to be required to be on.
+  set(CMAKE_CXX_EXTENSIONS ON)
```
@mathstuf Is requiring the `CMAKE_CXX_EXTENSIONS` a known limitation?

mathstuf replied:
CMake doesn't care about this in the relevant codepaths, so if it is, it's a compiler thing.

mordante replied:
I think it's a CMake issue, I've filed https://gitlab.kitware.com/cmake/cmake/-/issues/25916 with additional information.

hawkinsw commented on libcxx/docs/Modules.rst:
```diff
@@ -105,9 +103,17 @@ Users need to be able to build their own BMI files.
+    Clang 17 or later and CMake 3.26 or later.
+
+  * Use the installed modules. This requires Clang 18.1.2 or later and
+    a recent build of CMake. The CMake changes will be part CMake 3.30. This
```
```suggestion
    a recent build of CMake. The CMake changes will be part of CMake 3.30. This
```

hawkinsw commented on libcxx/docs/Modules.rst:
```diff
@@ -214,6 +219,62 @@ Building this project is done with the following steps, assuming the files
+
+CMake has added experimental support for importing the standard modules. This
+is available in the current nightly builds and will be part of the 3.30
+release. Currently CMake only supports importing the Standard modules in C++23
```
Do we want to write "Standard" or "standard"?

mordante replied:
The official wording uses `International Standard` so I tend to capitalize; that is if I don't forget to do it ;-)

hawkinsw replied:
I thought so -- I am terrible with that sort of thing :-)

hawkinsw reviewed (commented):
Thank you for adding this helpful documentation! I hope that my little bit of feedback is helpful.

mordante committed:
Update libcxx/docs/Modules.rst

Co-authored-by: Will Hawkins <whh8b@obs.cr>

mordante committed:
Updates patch

- Addresses review comment.
- Updates CMake status regarding disabling compiler extensions.

ldionne commented on libcxx/docs/Modules.rst:
```diff
@@ -214,6 +219,64 @@ Building this project is done with the following steps, assuming the files
+developers.
+
+The example uses the same ``main.cpp`` as above. It uses the following
+``CMakeLists.txt``
```
```suggestion
``CMakeLists.txt``:
```

ldionne commented on libcxx/docs/Modules.rst:
```diff
@@ -214,6 +219,64 @@ Building this project is done with the following steps, assuming the files
+  set(CMAKE_EXPERIMENTAL_CXX_IMPORT_STD "0e5b6991-d74f-4b3d-a41c-cf096e0b2508")
+  set(CMAKE_CXX_MODULE_STD ON)
+
+  project("module"
```
```suggestion
  project("example"
```

ldionne commented:
At this point, I think we should consider installing the modules source files by default. It will still be up to the vendors whether they want to ship them. Once vendors start installing them (even if they're still experimental), this would get the ball rolling for build systems to add support for modules. It's a bit of a chicken-and-egg problem, someone needs to decide to ship something that isn't 100% ready yet. I think is has to be us.
